pkg/plugins/generic/generic_plugin.go

```diff
@@ -69,31 +109,31 @@ func (p *GenericPlugin) OnNodeStateChange(new *sriovnetworkv1.SriovNetworkNodeSt
 	p.DesireState = new
 
 	needDrain = needDrainNode(new.Spec.Interfaces, new.Status.Interfaces)
-	needReboot = needRebootNode(new, &p.LoadVfioDriver, &p.LoadVirtioVdpaDriver)
+	needReboot = needRebootNode(new, p.DriverStateMap)
```

Comment:
Just want to add that what you did is definitely an improvement so please don't undo that :)

Reply:
Initially, the state is `unloaded`, then `needRebootNode` get's called which sets the state to `loading` if needed. Next, `Apply()` is called, which results in calling `p.HostManager.LoadKernelModule` and immediately setting to `loaded`. 

Without asking for any change, I'm just wondering if we can rid of this intermediate state? Why do have a tri-state thing. Can't we just have load + loaded without loading?

It seems to me that loading means "not loaded but need to load. Tried it an failed". Maybe the "need to load" part doesn't need to be cached to simplify the code?

Reply:
Maybe I can write a reduced version of needRebootNode that doesn't set the loading state at all, but just calls the tryEnableIommuInKernelArgs function in case vfio needs to be loaded, this is for knowing if the node has to be rebooted or not. This is to maintain compatibility with the existing code in daemon.go (drain and reboot flags are handled at least in two places, daemon.go and generic_plugin.go).

Reply:
That makes more sense to me. Ideally, we would want two functions, one to figure out if a reboot is needed, and then a separate one that actually sets the kargs. Currently, the "optimization" of setting the kargs and returning if it actually needs a reboot in the process makes the code unnecessary complex.

`nodeRebootNeeded` and `ensureKargsSet` could be the two functions. If we have those two functions, do we still need 3 states?

I realize you just wanted to add vhost-vdpa support and now we are refactoring the code, but if we just keep adding stuff, the code gets worse over time. Let's put in an active effort to improve the code before we make changes.

Reply:
Removed the `loading` state. Hope it is what you meant. Please review again

Reply:
yes! this looks better to me. I'd like the input of others too of course.

Reply:
IMO removing this tri-state is fine. 

one thing that bothers me, is that a function that just suppose to check for reboot actually does things to the system (modify kargs, writes config files etc...) however thats not related to this PR.